### PR TITLE
Add support for writing systems in Python

### LINKED ISSUE #790: Scripting: plugins that aren't written in C++
## Desired behavior

In the interest of making Gazebo more approachable and quicker to use, we could offer a scripting interface that lets users write plugins in Python (and potentially other languages too, like Ruby and Javascript).

The user could load their Python script through the SDF's `<plugin>` tag, just like they do with C++ libraries right now.

## Alternatives considered

Users could use ROS to interact with Gazebo using various scripting languages. We could also just offer a scripting API to Ignition Transport, so users can interact with simulation asynchronously.

These 2 approaches limit what the user can do though, because they're constrained to the transport API and must run async.

## Implementation suggestion

First we'll need to expose C++ APIs that are commonly used by plugins to scripts, such as:

* The [EntityComponentManager](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo5/include/ignition/gazebo/EntityComponentManager.hh)'s API so users can access and modify entities and components
* Component data types so that they can be read and manipulated. Components commonly include Math, SDFormat and Msgs types. 
* All the [System](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo5/include/ignition/gazebo/System.hh) callbacks need to be inherited by the Python plugin so that they're called at the correct time during simulation.

We can either add C APIs to all of these and then call those C functions from scripts, or use SWIG to generate APIs in other languages from C++. 

Finally, we'll need to support loading those scripts from the `<plugin>` tag.

## Additional context

Related to https://github.com/ignitionrobotics/ign-gazebo/issues/789

### PR DESCRIPTION
**Pull request by azeey**
# 🎉 New feature

Closes #790 

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: examples/scripts/python_api/systems/test_system.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from gz.sim8 import UpdateInfo, EntityComponentManager, Model
```
Comment (ahcorde) [inline examples/scripts/python_api/systems/test_system.py:1]: LICENSE

File: examples/worlds/python_system_loader.sdf [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+```
+  export PYTHONPATH=path_to_colcon_workspace/install/lib/python:$PYTHONPATH
+```
+-->
+<sdf version="1.6">
+  <world name="python_system">
+
+
+    <light type="directional" name="sun">
```
Comment (ahcorde) [inline examples/worlds/python_system_loader.sdf:23]:
```suggestion
  <world name="python_system">

    <light type="directional" name="sun">
```

File: src/ServerPrivate.hh [hunk 3]
```diff
@@ -22,11 +22,15 @@
 #include <gz/msgs/server_control.pb.h>
 #include <gz/msgs/stringmsg_v.pb.h>
 
+
```
Comment (ahcorde) [inline src/ServerPrivate.hh:25]: remove

File: src/systems/python_system_loader/PythonSystemLoader.cc [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+    gzdbg << "Searching for python module in:\n";
+    auto searchPaths = sys.attr("path");
+    // TODO (azeey) Improve formatting.
+    gzdbg << py::str(searchPaths).cast<std::string>() << std::endl;
+  }
+  catch (const pybind11::error_already_set &_err)
+  {
+    gzerr << "Error while loading module 'sys'\n"
+      << _err.what() << std::endl;
```
Comment (ahcorde) [inline src/systems/python_system_loader/PythonSystemLoader.cc:62]:
```suggestion
    gzerr << "Error while loading module 'sys'\n"
          << _err.what() << std::endl;
```

File: src/systems/python_system_loader/PythonSystemLoader.cc [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    return;
+  }
+
+  if (_method)
+  {
+    try
+    {
+      _method(std::forward<Args>(_args)...);
+    }
+    catch (const pybind11::error_already_set &_err)
```
Comment (Voldivh) [inline src/systems/python_system_loader/PythonSystemLoader.cc:131]:
```suggestion
    catch (const py::error_already_set &_err)
```

**Review by Voldivh — COMMENTED** [review]
Left a couple of minor comments. This looks good! Is it ready to try it out? @azeey

**Comment by azeey** [comment]
> Left a couple of minor comments. This looks good! Is it ready to try it out? @azeey

Yes, I have an example with some instructions.

File: src/systems/python_system_loader/PythonSystemLoader.cc [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    {
+      sys.attr("path").attr("append")(pluginPath);
+    }
+
+    gzdbg << "Searching for python module in:\n";
+    auto searchPaths = sys.attr("path");
+    // TODO (azeey) Improve formatting.
+    gzdbg << py::str(searchPaths).cast<std::string>() << std::endl;
+  }
+  catch (const pybind11::error_already_set &_err)
```
Comment (Voldivh) [inline src/systems/python_system_loader/PythonSystemLoader.cc:59]:
```suggestion
  catch (const py::error_already_set &_err)
```

File: src/systems/python_system_loader/PythonSystemLoader.cc [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    gzerr << "Error while loading module 'sys'\n"
+      << _err.what() << std::endl;
+    return;
+  }
+
+  try
+  {
+    this->pythonModule = py::module_::import(moduleName.c_str());
+  }
+  catch (const pybind11::error_already_set &_err)
```
Comment (Voldivh) [inline src/systems/python_system_loader/PythonSystemLoader.cc:70]:
```suggestion
  catch (const py::error_already_set &_err)
```

File: src/systems/python_system_loader/PythonSystemLoader.cc [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+  {
+    py::object getSystem = this->pythonModule.attr("get_system");
+    if (!getSystem)
+    {
+      gzerr << "Could not call 'get_system' on the module" << moduleName << "\n";
+      return;
+    }
+    this->pythonSystem = getSystem();
+  }
+  catch (const pybind11::error_already_set &_err)
```
Comment (Voldivh) [inline src/systems/python_system_loader/PythonSystemLoader.cc:87]:
```suggestion
  catch (const py::error_already_set &_err)
```

**Comment by azeey** [comment]
It seems that github automatically closed this PR because the branch it was targeting was deleted. I'm also not able to reopen it, so I'll create a new PR.

**Pull request closed without merging**